File: internal/git/git.go
package git

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type Worktree struct {
	Path   string
	Head   string
	Branch string
	Bare   bool
}

func runGit(args []string, cwd string) (string, error) {
	cmd := exec.Command("git", args...)
	if cwd != "" {
		cmd.Dir = cwd
	}
	out, err := cmd.CombinedOutput()
	output := strings.TrimSpace(string(out))
	if err != nil {
		return output, fmt.Errorf("git %s failed: %s", strings.Join(args, " "), output)
	}
	return output, nil
}

func ListWorktrees(cwd string) ([]Worktree, error) {
	out, err := runGit([]string{"worktree", "list", "--porcelain"}, cwd)
	if err != nil {
		return nil, err
	}
	if out == "" {
		return nil, nil
	}

	var worktrees []Worktree
	blocks := strings.Split(out, "\n\n")

	for _, block := range blocks {
		lines := strings.Split(strings.TrimSpace(block), "\n")
		if len(lines) == 0 || lines[0] == "" {
			continue
		}

		var wt Worktree
		for _, line := range lines {
			switch {
			case strings.HasPrefix(line, "worktree "):
				wt.Path = strings.TrimPrefix(line, "worktree ")
			case strings.HasPrefix(line, "HEAD "):
				wt.Head = strings.TrimPrefix(line, "HEAD ")
			case strings.HasPrefix(line, "branch "):
				wt.Branch = strings.TrimPrefix(line, "branch ")
			case line == "detached":
				wt.Branch = ""
			case line == "bare":
				wt.Bare = true
			}
		}
		if wt.Path != "" {
			worktrees = append(worktrees, wt)
		}
	}
	return worktrees, nil
}

func IsGitRoot(cwd string) bool {
	if cwd == "" {
		var err error
		cwd, err = os.Getwd()
		if err != nil {
			return false
		}
	}
	out, err := runGit([]string{"rev-parse", "--show-toplevel"}, cwd)
	if err != nil {
		return false
	}
	// Resolve symlinks for comparison (macOS /var -> /private/var)
	resolved, err := filepath.EvalSymlinks(cwd)
	if err != nil {
		return false
	}
	outResolved, err := filepath.EvalSymlinks(out)
	if err != nil {
		return false
	}
	return resolved == outResolved
}

func GetCurrentRef(cwd string) (string, error) {
	out, err := runGit([]string{"symbolic-ref", "--short", "HEAD"}, cwd)
	if err == nil {
		return out, nil
	}
	return runGit([]string{"rev-parse", "HEAD"}, cwd)
}

func GetHeadSha(cwd string) (string, error) {
	return runGit([]string{"rev-parse", "HEAD"}, cwd)
}

func HasUncommittedChanges(cwd string) (bool, error) {
	out, err := runGit([]string{"status", "--porcelain"}, cwd)
	if err != nil {
		return false, err
	}
	return len(out) > 0, nil
}

func StashChanges(cwd string) (bool, error) {
	out, err := runGit([]string{"stash", "-u"}, cwd)
	if err != nil {
		return false, err
	}
	return !strings.Contains(out, "No local changes to save"), nil
}

func PopStash(cwd string) (success bool, conflicted bool, err error) {
	_, err = runGit([]string{"stash", "pop"}, cwd)
	if err == nil {
		return true, false, nil
	}
	if strings.Contains(err.Error(), "CONFLICT") {
		return false, true, nil
	}
	return false, false, err
}

func CreateCheckpoint(cwd string) (string, error) {
	if _, err := runGit([]string{"add", "-A"}, cwd); err != nil {
		return "", err
	}
	if _, err := runGit([]string{"commit", "-m", "spotlight checkpoint"}, cwd); err != nil {
		return "", err
	}
	return runGit([]string{"rev-parse", "HEAD"}, cwd)
}

func AmendCheckpoint(cwd string) (string, error) {
	if _, err := runGit([]string{"add", "-A"}, cwd); err != nil {
		return "", err
	}
	if _, err := runGit([]string{"commit", "--amend", "--no-edit"}, cwd); err != nil {
		return "", err
	}
	return runGit([]string{"rev-parse", "HEAD"}, cwd)
}

func GetTreeSha(commitSha string, cwd string) (string, error) {
	return runGit([]string{"rev-parse", commitSha + "^{tree}"}, cwd)
}

func CheckoutCommit(sha string, cwd string) error {
	_, err := runGit([]string{"checkout", sha}, cwd)
	return err
}

func ForceCheckoutRef(ref string, cwd string) error {
	_, err := runGit([]string{"checkout", "--force", ref}, cwd)
	return err
}

func SoftReset(sha string, cwd string) error {
	_, err := runGit([]string{"reset", "--soft", sha}, cwd)
	return err
}

File: src/lib/git.ts
export interface Worktree {
  path: string;
  head: string;
  branch: string | null;
  bare: boolean;
}

function runGit(args: string[], cwd?: string): { stdout: string; stderr: string } {
  const result = Bun.spawnSync(["git", ...args], {
    cwd,
    stdout: "pipe",
    stderr: "pipe",
  });

  const stdout = result.stdout.toString().trim();
  const stderr = result.stderr.toString().trim();

  if (result.exitCode !== 0) {
    throw new Error(
      `git ${args.join(" ")} failed (exit ${result.exitCode}): ${stderr}`,
    );
  }

  return { stdout, stderr };
}

export function listWorktrees(cwd?: string): Worktree[] {
  const { stdout } = runGit(["worktree", "list", "--porcelain"], cwd);
  if (!stdout) return [];

  const blocks = stdout.split("\n\n");
  const worktrees: Worktree[] = [];

  for (const block of blocks) {
    const lines = block.trim().split("\n");
    if (lines.length === 0 || !lines[0]) continue;

    let path = "";
    let head = "";
    let branch: string | null = null;
    let bare = false;

    for (const line of lines) {
      if (line.startsWith("worktree ")) {
        path = line.slice("worktree ".length);
      } else if (line.startsWith("HEAD ")) {
        head = line.slice("HEAD ".length);
      } else if (line.startsWith("branch ")) {
        branch = line.slice("branch ".length);
      } else if (line === "detached") {
        branch = null;
      } else if (line === "bare") {
        bare = true;
      }
    }

    if (path) {
      worktrees.push({ path, head, branch, bare });
    }
  }

  return worktrees;
}

export function isGitRoot(cwd?: string): boolean {
  const { realpathSync } = require("node:fs");
  const dir = realpathSync(cwd ?? process.cwd());
  try {
    const { stdout } = runGit(["rev-parse", "--show-toplevel"], dir);
    return realpathSync(stdout) === dir;
  } catch {
    return false;
  }
}

export function getCurrentRef(cwd?: string): string {
  try {
    const { stdout } = runGit(["symbolic-ref", "--short", "HEAD"], cwd);
    return stdout;
  } catch {
    const { stdout } = runGit(["rev-parse", "HEAD"], cwd);
    return stdout;
  }
}

export function getHeadSha(cwd?: string): string {
  const { stdout } = runGit(["rev-parse", "HEAD"], cwd);
  return stdout;
}

export function hasUncommittedChanges(cwd?: string): boolean {
  const { stdout } = runGit(["status", "--porcelain"], cwd);
  return stdout.length > 0;
}

export function stashChanges(cwd?: string): boolean {
  const { stdout } = runGit(["stash", "-u"], cwd);
  return !stdout.includes("No local changes to save");
}

export function popStash(cwd?: string): { success: boolean; conflicted: boolean } {
  const result = Bun.spawnSync(["git", "stash", "pop"], {
    cwd,
    stdout: "pipe",
    stderr: "pipe",
  });

  const stderr = result.stderr.toString();

  if (result.exitCode === 0) {
    return { success: true, conflicted: false };
  }

  const conflicted = stderr.includes("CONFLICT");
  return { success: false, conflicted };
}

export function createCheckpoint(cwd: string): string {
  runGit(["add", "-A"], cwd);
  runGit(["commit", "-m", "spotlight checkpoint"], cwd);
  const { stdout } = runGit(["rev-parse", "HEAD"], cwd);
  return stdout;
}

export function amendCheckpoint(cwd: string): string {
  runGit(["add", "-A"], cwd);
  runGit(["commit", "--amend", "--no-edit"], cwd);
  const { stdout } = runGit(["rev-parse", "HEAD"], cwd);
  return stdout;
}

export function getTreeSha(commitSha: string, cwd?: string): string {
  const { stdout } = runGit(["rev-parse", `${commitSha}^{tree}`], cwd);
  return stdout;
}

export function checkoutCommit(sha: string, cwd?: string): void {
  runGit(["checkout", sha], cwd);
}

export function forceCheckoutRef(ref: string, cwd?: string): void {
  runGit(["checkout", "--force", ref], cwd);
}

export function softReset(sha: string, cwd: string): void {
  runGit(["reset", "--soft", sha], cwd);
}
